pkg/providers/vmware/provider.go

```diff
@@ -0,0 +1,422 @@
+func (r *VmwareProvider) FindTemplate() (*oapiv1.Template, error) {
+	vm, err := r.getVmProperties()
+	if err != nil {
+		return nil, err
+	}
+	return r.templateFinder.FindTemplate(vm)
+}
+
+// ProcessTemplate uses the Openshift API to process a template
+func (r *VmwareProvider) ProcessTemplate(template *oapiv1.Template, vmName *string, namespace string) (*v1.VirtualMachine, error) {
+	vm, err := r.templateHandler.ProcessTemplate(template, vmName, namespace)
+	if err != nil {
+		return nil, err
+	}
+	vmProperties, err := r.getVmProperties()
+	if err != nil {
+		return nil, err
+	}
+	labels, annotations, err := r.templateFinder.GetMetadata(template, vmProperties)
+	if err != nil {
+		return nil, err
+	}
+	updateLabels(vm, labels)
+	updateAnnotations(vm, annotations)
+	return vm, nil
+}
+
+// PrepareResourceMapping merges the external resource mapping with the mapping provided in the VirtualMachineImport spec
+func (r *VmwareProvider) PrepareResourceMapping(externalResourceMapping *v1beta1.ResourceMappingSpec, vmiSpec v1beta1.VirtualMachineImportSourceSpec) {
+	r.resourceMapping = mappings.MergeMappings(externalResourceMapping, vmiSpec.Vmware.Mappings)
+}
+
+// LoadVM fetches the source VM.
+func (r *VmwareProvider) LoadVM(sourceSpec v1beta1.VirtualMachineImportSourceSpec) error {
+	vmwareClient, err := r.getClient()
+	if err != nil {
+		return err
+	}
+	vm, err := vmwareClient.GetVM(sourceSpec.Vmware.VM.ID, sourceSpec.Vmware.VM.Name, nil, nil)
+	if err != nil {
+		return err
+	}
+	r.vm = vm.(*object.VirtualMachine)
+	return nil
+}
+
+// GetVMName gets the name of the source VM
+func (r *VmwareProvider) GetVMName() (string, error) {
+	vm, err := r.getVM()
+	if err != nil {
+		return "", err
+	}
+	return vm.Name(), nil
+}
+
+// GetVMStatus gets the power status of the source VM
+func (r *VmwareProvider) GetVMStatus() (provider.VMStatus, error) {
+	vmProperties, err := r.getVmProperties()
+	if err != nil {
+		return "", err
+	}
+
+	poweredOn := types.VirtualMachinePowerStatePoweredOn
+	poweredOff := types.VirtualMachinePowerStatePoweredOff
+
+	switch vmProperties.Runtime.PowerState {
+	case poweredOn:
+		return provider.VMStatusUp, nil
+	case poweredOff:
+		return provider.VMStatusDown, nil
+	}
+
+	return "", fmt.Errorf("VM doesn't have a legal status. Allowed statuses: [%v, %v]", poweredOn, poweredOff)
+}
+
+// StartVM powers on the source VM.
+func (r *VmwareProvider) StartVM() error {
+	vmwareClient, err := r.getClient()
+	if err != nil {
+		return err
+	}
+	vm, err := r.getVM()
+	if err != nil {
+		return err
+	}
+	return vmwareClient.StartVM(vm.Reference().Value)
+}
+
+// StartVM powers off the source VM.
+func (r *VmwareProvider) StopVM(cr *v1beta1.VirtualMachineImport, client client.Client) error {
+	vmwareClient, err := r.getClient()
+	if err != nil {
+		return err
+	}
+	vm, err := r.getVM()
+	if err != nil {
+		return err
+	}
+	err = vmwareClient.StopVM(vm.Reference().Value)
```

Comment:
Provider's `StopVM` method is called in every reconciliation and `vmwareClient.StopVM` always retrieves VM from VMware, which we already have in provider. In single migration one vm import is reconciled many times, and when multiplied by number of VMs in mass migration it can give really big number of requests to VMware API. Is it possible to check the power status here?

Reply:
`vmwareClient.StopVM` uses `getByMoRef` to get the VM, which does not involve an API call-- it just constructs a VM object from the reference. `vm.PowerState()` does however involve a vSphere API call each time. I'll see about using the vmProperties that the provider has already collected to check the power state while avoiding the API call.

Reply:
:+1: